### DISCUSSION TITLE
Moving in a 3D space

### DISCUSSION THREAD

**Discussion by olivierpls** [mholtkamp/octave#13] [Q&A] [answered]
Hi ! In my free time I have fun using this superb game engine to develop a small game which will be ported to GameCube in the end. 

However, Iwant to be able to move a character in a 3D world with a third person view. 
I have already managed to link a camera to my character, and also managed to move my character on the 3 axes using the keyboard. 
I also added a rotation of the character so that I can look left and right, and now I'm facing a problem. 

When I use rotation to look in another direction and try to move my character forward, it does not move in the direction I am looking, but moves on all 3 axes as originally without the rotation. 
I thought I understood that I had to use vectors **(GetForwardVector(), GetUpVector(), GetRightVector())** depending on the direction in which my character is facing, but I can't do it. 

I guess I'm not doing it the right way, since at the moment I'm trying to move the character, when I should rather be moving the map? But being a beginner and doing this in my free time as a hobby, I unfortunately don't know how to use the right functions to move my character.

_Infos about the code_

Default Scene name : **SC_Default**
Caracter scene name : **SC_Bird**
Obstacle scene name : **SC_Tuyau**

GetPosition() stored in variable **newPos**
GetWorldPosition() stored in variable **worldPos**
GetRotation() stored in variable **pouleRot**

Thank you in advance for your help

[3D_World.zip](https://github.com/user-attachments/files/16398043/3D_World.zip)

**Comment by olivierpls** [comment]
I managed to solve my problem! 

To do this, I constantly retrieve the vector value of my current position with **GetPosition()**, 
then when I press the keyboard to move forward, I retrieve the vector value of the position I'm looking at with **GetForwardVector()**, 
and then I multiply this vector by the speed (chosen beforehand), 
I multiply by **deltaTime**, then I add everything to the value of my position previously retrieved with **GetPosition()**, and VOILA. 

Now my character moves well in the direction I point him. Same procedure for moving from left to right but using **GetRightVector()** instead of **GetForwardVector()**.

Now I have a new problem... I can move forward, backward, left, right, but, when I want for example to move at an angle (forward and left at the same time), I moves either only forward or only to the left, depending on which keyboard key was last pressed.

And I admit that I don't know how to fix this.


Part of the code that I use to move my caracter : 

```Lua
local newPos = self.poule:GetPosition()

-- Avancer
    if (Input.IsKeyDown(Key.Z)) then
        directionPoule = self.poule:GetForwardVector()
        directionPoule = -directionPoule * self.vitesse * deltaTime + newPos
    end
    -- Reculer
    if (Input.IsKeyDown(Key.S)) then
        directionPoule = self.poule:GetForwardVector()
        directionPoule = directionPoule * self.vitesse * deltaTime + newPos
    end
    -- Aller à gauche
    if (Input.IsKeyDown(Key.Q)) then
        directionPoule = self.poule:GetRightVector()
        directionPoule = directionPoule * self.vitesse * deltaTime  + newPos
    end
    -- Aller à droite
    if (Input.IsKeyDown(Key.D)) then
        directionPoule = self.poule:GetRightVector()
        directionPoule = -directionPoule * self.vitesse * deltaTime  + newPos
    end

self.poule:SetPosition(directionPoule)
```

**Reply by mholtkamp** [reply, marked as answer]
I'm glad you were able to solve your original problem!

I think you are having trouble moving at an angle because you are overwriting directionPoule. If Z is pressed, then you set directionPoule, but if Q is also pressed then you are resetting directionPoule before you ever set the new position.

Maybe something like this would work:
```lua
    local forward = self.poule:GetForwardVector()
    local right = self.poule:GetRightVector()

    directionPoule = Vec(0, 0, 0) 

    -- Avancer
    if (Input.IsKeyDown(Key.Z)) then
        directionPoule = directionPoule - forward
    end
    -- Reculer
    if (Input.IsKeyDown(Key.S)) then
        directionPoule = directionPoule + forward
    end
    -- Aller à gauche
    if (Input.IsKeyDown(Key.Q)) then
        directionPoule = directionPoule + right
    end
    -- Aller à droite
    if (Input.IsKeyDown(Key.D)) then
        directionPoule = directionPoule - right
    end

    newPos = newPos + (directionPoule * self.vitesse * deltaTime)
```

I didn't actually test the above code, but the idea is that you can accumulate multiple direction vectors into directionPoule, and then use the combination when updating newPos.

**Reply by olivierpls** [reply]
Hi !
Thanks for the help, it allowed me to move forward and understand how it works a little better :)
I followed your advice and even implemented a functionality to control the character with a controller! 
Attached the Zip file.

Fully functional now, I think I can close the thread? Next problem, managing collisions lol
[V5.zip](https://github.com/user-attachments/files/16402501/V5.zip)

**Reply by mholtkamp** [reply]
For collisions, you will probably want to use Primitive3D's SweepToPosition() function. You can pass in a world-space position and then the engine will move the node as close as it can to the target position, but will stop if it collides with something. SweepToPosition() can only be called on nodes with convex collision like Sphere3D, Box3D, and Capsule3D, but it will collide against anything with collision (even a non-convex triangle mesh)

Once you get some form of basic collision detection working, you can try getting smoother movement by using two consecutive SweepToPosition() calls. Here's an example:

```lua
function Bomber:Move(axes, deltaTime)

    -- Sweep along velocity
    local velocity = self.velocity * axes
    self.velocity = self.velocity - velocity
    local nodePos = self:GetWorldPosition()
    local endPos = nodePos + velocity * deltaTime
    local sweepRes = self:SweepToPosition(endPos)

    if (sweepRes.hitNode) then

        -- Uncomment to debug the hit normal.
        --Renderer.AddDebugLine(sweepRes.hitPosition, sweepRes.hitPosition + sweepRes.hitNormal * 1.0, Vec(0, 1, 0, 1), 3.0)

        -- Update to new position
        nodePos = self:GetWorldPosition()

        -- Cancel out velocity along normal and perform a second sweep
        velocity = velocity - (sweepRes.hitNormal * Vector.Dot(velocity, sweepRes.hitNormal))
        endPos = nodePos + velocity * deltaTime
        sweepRes = self:SweepToPosition(endPos)
    end

    self.velocity = self.velocity + velocity

end
```

And then I call it like this:
```lua
        -- Move in Y direction first
        self:Move(Vec(0, 1, 0), deltaTime)

        -- Move in X/Z plane next (for smoother movement along obstacles)
        self:Move(Vec(1, 0, 1), deltaTime)
```


The full file is located here:
https://github.com/mholtkamp/octo-bombers/blob/main/Bomber/Scripts/Bomber.lua

**Reply by olivierpls** [reply]
A very big thank you to you for your precious help. I can now move my character in a 3-dimensional world with a gamepad, and collisions are now functional. 

I'm so happy to have managed to get this far without having much knowledge of the .lua language, but thanks to your videos and your advice, I can finally realize my childhood wish and develop a very small game for my favorite console , the GameCube.

Now I just have to manage to move the camera to look up and down without my character subsequently moving in the direction of where I am looking on the Y axis (up and down). But hey, that will be for next time lol.

Attached is the .zip file of my work so far, to be used by anyone who wants it.
[V6.zip](https://github.com/user-attachments/files/16432477/V6.zip)

**Comment by olivierpls** [comment]
Hi, quick update on my progress.

So, compared to last time I managed to use the RotateAround() function which allowed me to move the camera up and down around my character in order to finally be able to look up and/or down without that my character doesn't fly towards the direction I'm looking lol.

I also tried to model terrain in Blender, I added a texture and imported it into my default scene. 
The collision with the ground is well detected and my character manages to move correctly following the reliefs. 

I also added a function that allows you to manage the double jump, which counts the number of jumps while the character is in the air, and once the chosen number is reached (2 in my case), prevent the character from jumping again as long as it has not touched the ground and/or a surface on which collision is activated.

I'm moving forward slowly and little by little I'm discovering the joys of developing a video game and I find it very exciting! 
Thanks again for this amazing video game engine.

Attached is the .zip file of my work so far.
[V8.zip](https://github.com/user-attachments/files/16460879/V8.zip)

**Reply by mholtkamp** [reply]
Wow, that's amazing progress! I'm looking forward to seeing what you can make with the engine. I apologize in advance for any crashes you experience and also the lack of documentation. Feel free to keep asking questions if you get stuck or experience any other issues. Good luck brave soul and thank you for testing out my engine :) !

**Comment by olivierpls** [comment]
Thanks for the encouragement. 
I am now facing a new problem. 

I created a scene called 'SC_Assets' in which I added several nodes (a cube, a ball, a tree, etc.) and I am trying to make them appear in my SC_Default when I press for example the X button to make a ball appear, the Y button for a tree etc.. 

Until now I have managed to import a complete scene with all the nodes , but I can't import only one node that I have chosen in the scene.

I used the LoadAsset() functions to load the scene I want, and Instantiate() to link it to my SC_Default, and then AddChild to make my scene appear at the desired location.

I guess my mistake is because I'm using Instantiate() to bind the whole scene and not another function to pick just one element from a scene, but I don't know which function to use to achieve the desired result.

**Reply by mholtkamp** [reply]
There currently isn't functionality to instantiate only a subset of a scene, but if you are only spawning a simple mesh you could do something like this:

```lua
            local staticMesh = nil
            if (Input.IsGamepadPressed(Gamepad.Y)) then
                staticMesh = LoadAsset('SM_Tree')
            elseif (Input.IsGamepadPressed(Gamepad.X)) then
                staticMesh = LoadAsset('SM_Ball')
            end

            if (staticMesh ~= nil) then
                local meshNode = Node.Construct('StaticMesh3D')
                meshNode:SetStaticMesh(staticMesh)
                self:AddChild(meshNode)

                -- Then place it where you want
                meshNode:SetPosition(Vec(1, 2, 3))
            end
```

But if the Tree and Ball are very different scenes (like for instance you have physics enabled on the ball with a certain mass, and restitution value) then maybe it's better to save them as separate scenes and do the same thing as above ^ but using scene:Instantiate() instead of Node.Construct()

If you already did a lot of work in your combined scene, you can click on a specific sub-node in the scene, right click in the asset browser somewhere and select "Capture Selected Node" and that will create a separate scene from the node you have selected.
